### PR TITLE
Add new Snap ML estimators

### PR DESCRIPTION
**Pull request by tdoublep**
This pull request introduces 3 more estimators from Snap ML:
```
DecisionTreeRegressor
BoostingMachineRegressor
BoostingMachineClassifier
```
I've also carefully reviewed the existing 3 estimators and made sure their defaults align with the latest version of Snap ML and also that the tunable ranges align with what we are using in AutoAI.

### CODE REVIEW

**Comment by tdoublep** [comment]
@hirzel please take a look when you have time. don't seem to have permissions to add reviewers.

File: lale/lib/snapml/decision_tree_regressor.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                        },
+                    ],
+                    "default": 1,
+                    "description": "The minimum number of samples required to be at a leaf node.",
+                },
+                "max_features": {
+                    "anyOf": [
+                        {
+                            "type": "integer",
+                            "minimum": 1,
```
Comment (hirzel) [inline lale/lib/snapml/decision_tree_regressor.py:137]: Should add a laleMax here for checking against the number of columns in the dataset, like this: https://github.com/IBM/lale/blob/5013699e88add4e79000c2f889c2c8450e7795e0/lale/lib/sklearn/decision_tree_regressor.py#L142

**Review by hirzel — APPROVED** [review]
Looks very good overall! I've added some minor comments.

File: lale/lib/snapml/boosting_machine_regressor.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+from typing import TYPE_CHECKING
+
+try:
+    import snapml  # type: ignore
+
+    snapml_installed = True
+except ImportError:
+    snapml_installed = False
+    if TYPE_CHECKING:
```
[the remark this answers is not in the document]
Comment (tdoublep) [reply lale/lib/snapml/boosting_machine_regressor.py]: I've removed the second import, I agree it doesn't make sense.

File: lale/lib/snapml/boosting_machine_regressor.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+    def fit(self, X, y, **fit_params):
+        X = lale.datasets.data_schemas.strip_schema(X)
+        y = lale.datasets.data_schemas.strip_schema(y)
+        self._wrapped_model.fit(X, y, **fit_params)
+        return self
+
+    def predict(self, X, **predict_params):
+        X = lale.datasets.data_schemas.strip_schema(X)
+        if predict_params is None:
```
[the remark this answers is not in the document]
Comment (tdoublep) [reply lale/lib/snapml/boosting_machine_regressor.py]: Done.

File: lale/lib/snapml/boosting_machine_regressor.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+            "anyOf": [{"type": "array", "items": {"type": "number"}},],
+        },
+        "sample_weight": {
+            "anyOf": [
+                {"type": "array", "items": {"type": "number"}},
+                {"enum": [None], "description": "Samples are equally weighted."},
+            ],
+            "description": "Sample weights.",
+        },
+        "X_val": {
```
[the remark this answers is not in the document]
Comment (tdoublep) [reply lale/lib/snapml/boosting_machine_regressor.py:297]: Yes I see. I've also explicitly set the default to None -is this needed?

File: lale/lib/snapml/decision_tree_regressor.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                        },
+                    ],
+                    "default": None,
+                    "description": "The maximum depth of the tree.",
+                },
+                "min_samples_leaf": {
+                    "anyOf": [
+                        {
+                            "type": "integer",
+                            "minimum": 1,
```
[the remark this answers is not in the document]
Comment (tdoublep) [reply lale/lib/snapml/decision_tree_regressor.py:113]: Done

File: lale/lib/snapml/decision_tree_regressor.py (hunk 1 above)
Comment (tdoublep) [reply lale/lib/snapml/decision_tree_regressor.py:133]: Done

File: lale/lib/snapml/boosting_machine_regressor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hirzel) [reply lale/lib/snapml/boosting_machine_regressor.py:299]: Setting the default explicitly makes the auto-generated documentation on lale.readthedocs.io more clear.

[the remark this answers is not in the document]
Comment (tdoublep) [reply lale/lib/snapml/boosting_machine_regressor.py:296]: ok - makes sense! thx

**Pull request merged**